libraries/rush-lib/src/utilities/RushConnect.ts

```diff
@@ -0,0 +1,477 @@
+import {
+  AlreadyExistsBehavior,
+  Async,
+  FileConstants,
+  FileSystem,
+  JsonFile,
+  type INodePackageJson,
+  type IPackageJsonDependencyTable
+} from '@rushstack/node-core-library';
+import { type DependencyPath, depPathToFilename } from '@pnpm/dependency-path';
+import { PackageExtractor } from '@rushstack/package-extractor';
+import { pnpmSyncUpdateFileAsync, pnpmSyncCopyAsync, type ILogMessageCallbackOptions } from 'pnpm-sync-lib';
+import { parse } from '@pnpm/dependency-path';
+import * as semver from 'semver';
+
+import type { RushConfiguration } from '../api/RushConfiguration';
+import type { RushConfigurationProject } from '../api/RushConfigurationProject';
+import { RushConstants } from '../logic/RushConstants';
+import { PnpmSyncUtilities } from './PnpmSyncUtilities';
+
+class RushConnectError extends Error {
+  public constructor(message: string) {
+    super(message);
+    this.name = 'RushConnectError';
+  }
+}
+
+enum LinkType {
+  LinkPackage = 'LinkPackage',
+  BridgePackage = 'BridgePackage'
+}
+
+interface IRushLinkFileState {
+  [subspaceName: string]: {
+    linkedPackagePath: string;
+    linkedPackageName: string;
+    linkType: LinkType;
+  }[];
+}
+
+interface ILinkedPackageInfo {
+  packageName: string;
+  linkedPackageNodeModulesPath: string;
+  externalDependencies: string[];
+  workspaceDependencies: string[];
+  peerDependencies: IPackageJsonDependencyTable;
+}
+
+interface IConsumerPackageInfo {
+  consumerPackageNodeModulesPath: string;
+  consumerPackagePnpmDependenciesFolderPath: string;
+  consumerSubspaceName: string;
+}
+
+interface IRushLinkOptions {
+  rushLinkStateFilePath: string;
+  rushLinkState: IRushLinkFileState | undefined;
+}
+
+export class RushConnect {
+  public readonly rushLinkState: IRushLinkFileState | undefined;
+
+  private readonly _terminal: ITerminal;
+  private readonly _rushLinkStateFilePath: string;
+
+  public constructor(options: IRushLinkOptions) {
+    this._terminal = new Terminal(new ConsoleTerminalProvider());
+
+    this._rushLinkStateFilePath = options.rushLinkStateFilePath;
+    this.rushLinkState = options.rushLinkState;
+  }
+
+  private async _hardLinkToLinkedPackageAsync(
+    sourcePath: string,
+    targetFolder: string,
+    lockfileId: string
+  ): Promise<void> {
+    const logMessageCallback = (logMessageOptions: ILogMessageCallbackOptions): void => {
+      PnpmSyncUtilities.processLogMessage(logMessageOptions, this._terminal);
+    };
+    await pnpmSyncUpdateFileAsync({
+      sourceProjectFolder: sourcePath,
+      targetFolders: [targetFolder],
+      lockfileId,
+      logMessageCallback
+    });
+    const pnpmSyncJsonPath: string = path.resolve(
+      sourcePath,
+      RushConstants.nodeModulesFolderName,
+      '.pnpm-sync.json'
+    );
+    await pnpmSyncCopyAsync({
+      pnpmSyncJsonPath,
+      ensureFolderAsync: FileSystem.ensureFolderAsync,
```

Comment:
Bind this to `FileSystem`. We may be changing semantics here at some point.

```suggestion
      ensureFolderAsync: FileSystem.ensureFolderAsync.bind(FileSystem),
```